include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -176,9 +176,11 @@ class Ilu : public EnableLinOp<
         set_cache_to(b);
         if (!ReverseApply) {
             l_solver_->apply(b, cache_.intermediate.get());
+            x->copy_from(cache_.intermediate.get());
```

Comment:
I think we copy b to cache to make sure the cache is not random filled before.
Should the cache should be get data from x not b here?
first apply will use x as initial guess
second apply will use cache as initial guess

Reply:
I'm not sure that I completely understand the question, but I'm fairly certain that the data from x should not be used since it can be any vector, e.g. the krylo vector from a previous iteration. This doesn't matter for exact triangular solves since they overwrite x with the solution anyway. However, when using iterative triangular solves, this causes problems. The first apply solves:
Ly = b
By setting the cache to b, we use the right hand side as initial guess. Now the second apply solves
Ux = y
and by copying x from the cache, we again use the right hand side as initial guess.

Reply:
But I agree with @yhmtsai that `x` might be a better choice because, in the case of the solving `Ly=b` you are kind of hard-coding the init guess to `b`. But if you set the cache to `x`, would it not be possible for the function calling the preconditioner apply to set the initial guess and if you need it then you can set it to `b`, right ?

Reply:
I see what you mean. However, using the data from x as initial guess for the first solve introduces the same nonlinearity as before, only in a different spot. This way, we would have to ensure linearity of the preconditioner, a "linear operator", from outside the preconditioner before calling it, by putting adequate restrictions to x. 
I checked and tried GMRES with the following configurations (without writing extra ode in GMRESpreparing the data in x):
- only `set_cache_to(b)`: this was the initial version, and gives errors with iterative triangular solves
- only `set_cache_to(x)`: also gives errors in the result
- `set_cache_to(x)` and `x->copy_from(cache_.intermediate.get())`: also gives errors in the result
- `set_cache_to(b)` and `x->copy_from(cache_.intermediate.get())`: This is the only configuration for which the result GMRES computes is correct when using a few block Jacobi sweeps as triangular solves in the preconditioner.

Reply:
I typed up an explanation, hopefully this clarifies:
![explanation](https://user-images.githubusercontent.com/46599674/80190548-c624f480-8614-11ea-8635-0ac07202cfcc.png)